receiver/splunkenterprisereceiver/scraper.go

```diff
@@ -2,3 +2,199 @@
+
+import (
+	"context"
+	"encoding/json"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/splunkenterprisereceiver/internal/metadata"
+)
+
+var (
+	errMaxSearchWaitTimeExceeded = errors.New("Maximum search wait time exceeded for metric")
+)
+
+type splunkScraper struct {
+	splunkClient *splunkEntClient
+	settings     component.TelemetrySettings
+	conf         *Config
+	mb           *metadata.MetricsBuilder
+}
+
+func newSplunkMetricsScraper(params receiver.CreateSettings, cfg *Config) splunkScraper {
+	return splunkScraper{
+		settings: params.TelemetrySettings,
+		conf:     cfg,
+		mb:       metadata.NewMetricsBuilder(cfg.MetricsBuilderConfig, params),
+	}
+}
+
+// Create a client instance and add to the splunkScraper
+func (s *splunkScraper) start(_ context.Context, _ component.Host) (err error) {
+	c := newSplunkEntClient(s.conf)
+	s.splunkClient = &c
+	return nil
+}
+
+// The big one: Describes how all scraping tasks should be performed. Part of the scraper interface
+func (s *splunkScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	errs := &scrapererror.ScrapeErrors{}
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	s.scrapeLicenseUsageByIndex(ctx, now, errs)
+	s.scrapeIndexThroughput(ctx, now, errs)
+	return s.mb.Emit(), errs.Combine()
+}
+
+// Each metric has its own scrape function associated with it
+func (s *splunkScraper) scrapeLicenseUsageByIndex(ctx context.Context, now pcommon.Timestamp, errs *scrapererror.ScrapeErrors) {
+	var sr searchResponse
+	// Because we have to utilize network resources for each KPI we should check that each metrics
+	// is enabled before proceeding
+	if !s.conf.MetricsBuilderConfig.Metrics.SplunkLicenseIndexUsage.Enabled {
+		return
+	}
+
+	sr = searchResponse{
+		search: searchDict[`SplunkLicenseIndexUsageSearch`],
+	}
+
+	start := time.Now()
+	req, err := s.splunkClient.createRequest(ctx, &sr)
+	if err != nil {
+		errs.Add(err)
+	}
+
+	res, err := s.splunkClient.makeRequest(req)
+	if err != nil {
+		errs.Add(err)
+		return
+	}
+
+	err = unmarshallSearchReq(res, &sr)
+	if err != nil {
+		errs.Add(err)
+	}
+	res.Body.Close()
+
+	for ok := true; ok; ok = (sr.Return == 204) {
+		req, err = s.splunkClient.createRequest(ctx, &sr)
+		if err != nil {
+			errs.Add(err)
+		}
+
+		res, err = s.splunkClient.makeRequest(req)
+		if err != nil {
+			errs.Add(err)
+			res.Body.Close()
+			return
+		}
+
+		// if its a 204 the body will be empty because we are still waiting on search results
+
+		err = unmarshallSearchReq(res, &sr)
+		if err != nil {
+			errs.Add(err)
+		}
+		res.Body.Close()
+
+		if sr.Return == 204 {
+			time.Sleep(2 * time.Second)
+			continue
+		}
```

Comment:
This `continue` statement blocks the subsequent check of `MaxSearchWaitTime`, so the execution potentially can span way beyond the timeout.
```suggestion
		if sr.Return != 204 {
			break
		}
		time.Sleep(2 * time.Second)
```